### DISCUSSION TITLE
C# Markup - Null reference passing a BindableViewModel to a constructor of a UserControl

### DISCUSSION THREAD

**Discussion by mcNets** [unoplatform/uno#16850] [Q&A] [answered]
I'm trying out the new MVUX / C# Markup features and I'm having some issues working with UserControls and BindableViewModels. I'm not using frame navigation; instead, I use a ShellControl (UserControl) and dynamically change its content at runtime. The content of ShellControl changes based on the CurrentPage property of my NavigationService.

Here is my current MainWindow:

```csharp
internal class MainWindow : Window
{
    private BindableOsmoViewModel _osmoViewModel;
    public MainWindow()
    {
        _osmoViewModel = new BindableOsmoViewModel();
        Content = new ShellControl(_osmoViewModel);
    }
}
```

This is a simplified version of my ShellControl:

```csharp
internal class ShellControl : UserControl
{
    public ShellControl(BindableOsmoViewModel viewModel)
    {
        this.DataContext(viewModel, (control, vm) => 
            control.Content(
                new Border()
                    .Background(OsmoResources.BrushPageBackground)
                    .Child(
                        new OsmoControl(viewModel)
                            .CurrentPage(() => vm.NavService.CurrentPage))));
    }
}
```

And a simplified version of OsmoControl:

```csharp
internal class OsmoControl : UserControl
{
    private readonly Border _border;
    private readonly BindableOsmoViewModel _viewModel;

    public OsmoControl(BindableOsmoViewModel viewModel)
    {
        this._viewModel = viewModel;

        _border = new Border()
                    .HorizontalAlignment(HorizontalAlignment.Center)
                    .VerticalAlignment(VerticalAlignment.Center);

        Content = _border;
    }
}

// There is a CurrentPage dependency property that sets the border's content using a ContentFactory class.
```

The question is, if I pass `vm` to the constructor of OsmoControl, I get a null reference, however I can use `vm.NavService.CurrentPage` in the binding.

Why do I get a null value if I use `vm` in the OsmoControl's constructor?

In your [Calendar example](https://platform.uno/docs/articles/external/workshops/simple-calc/modules/MVUX-CSharp/03-Creating%20the%20Layout/README.html), you use `vm` in Header(), Output() and Keypad() controls.

**Comment by francoistanguay** [comment, marked as answer]
Instead of injecting your ViewModel in OsmoControl, just set the DataContext of the OsmoControl in the Shell, most probably using data binding, and then read this datacontext in the OsmoControl using this.DataContext(...). There's an overload where you can specify the type without specifying an instance. It will assume it is set in the parent somewhere.

Shell
  new OsmoControl().DataContext(() => ...)
  
OsmoControl
  this.DataContext<BindableOsmoViewModel>(vm => ...)
  
 
 Is it clear enough?
 
 cc @dr1rrb

**Reply by mcNets** [reply]
Thanks @francoistanguay, I'm testing it and I was just curious about this behaviour.

**Reply by mcNets** [reply]
Hi @francoistanguay,

I think it doesn't work with UserControls. Because the DataContext is set after the control has been created, and in your samples, you are using statically created UIElements.

I added a minimal example to illustrate what I mean. In this case, the bindings should be set, for example, in `OnDataContextChanged() `or by adding dependency properties that can be bound in `MainPage`.

What do you think?

MainPage:
```csharp
public sealed partial class MainPage : Page
{
    public MainPage()
    {
        this
            .Background(ThemeResource.Get<Brush>("ApplicationPageBackgroundThemeBrush"))
            .DataContext(new BindableMainModel(), (page, vm) => page
                .Content(
                    new StackPanel()
                        .VerticalAlignment(VerticalAlignment.Center)
                        .HorizontalAlignment(HorizontalAlignment.Center)
                        .Children(
                            new MyControl()
                                .DataContext(() => vm)
                        )));
    }
}
```

MyControl:
```csharp
internal partial class MyControl : UserControl
{
    public MyControl()
    {
        this.DataContext<BindableMainModel>((control, vm) =>
        {
            control.Content =
                new StackPanel()
                    .VerticalAlignment(VerticalAlignment.Center)
                    .HorizontalAlignment(HorizontalAlignment.Center)
                    .Children(
                        new StackPanel()
                            .Orientation(Orientation.Vertical)
                            .Spacing(10)
                            .Children(
                                new TextBlock()
                                    .Text(() => vm.Machine.IdName),

                                new Button()
                                    .Content("Click me!")
                                    .Command(() => vm.GetMachineCommand)
                            )
                    );
        });
    }
}
```

[UnoMvux1.zip](https://github.com/unoplatform/uno/files/15498925/UnoMvux1.zip)
